core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala

```diff
@@ -115,6 +115,7 @@ class ExternalAppendOnlyMap[K, V, C](
   private val keyComparator = new HashComparator[K]
   private val ser = serializer.newInstance()
 
+  @volatile private var isReadingIterator: Boolean = false
```

Comment:
Yeah, alternatively we can remove the assert and check if `readingIterator` is null or not. I just want to keep the original behavior.